### ISSUE TITLE
failed: (Exit 2): cl.exe failed: error executing command，Bazel=5.0.0, tf=2.8.0,MSVC2019 

### ISSUE THREAD

**Issue by Leemy9612** [tensorflow/tensorflow#54578]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):win11
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):build it from the source
- TensorFlow version:2.8.0
- Python version:3.10 && 3.8.8
- Installed using virtualenv? pip? conda?:
- Bazel version (if compiling from source):5.0.0 && 4.2.1
- GCC/Compiler version (if compiling from source):MSVC2019
- CUDA/cuDNN version:No
- GPU model and memory:No

**Describe the problem**
when I use bazel to build the tf2.8, the blow error happened.
`failed: (Exit 2): cl.exe failed: error executing command`

**Provide the exact sequence of commands / steps that you executed before running into the problem**
`F:\tensorflow>python ./configure.py`
You have bazel 5.0.0 installed.
Please specify the location of python. [Default is D:\JetBrain\python\python.exe]:

Found possible Python library paths:
D:\JetBrain\python\lib\site-packages
Please input the desired Python library path to use. Default is [D:\JetBrain\python\lib\site-packages]

Do you wish to build TensorFlow with ROCm support? [y/N]: n
No ROCm support will be enabled for TensorFlow.

Do you wish to build TensorFlow with CUDA support? [y/N]: n
No CUDA support will be enabled for TensorFlow.

Please specify optimization flags to use during compilation when bazel option "--config=opt" is specified [Default is /arch:AVX]:

Would you like to override eigen strong inline for some C++ compilation to reduce the compilation time? [Y/n]: y
Eigen strong inline overridden.

Would you like to interactively configure ./WORKSPACE for Android builds? [y/N]: n
Not configuring the WORKSPACE for Android builds.

Preconfigured Bazel build configs. You can use any of the below by adding "--config=<>" to your build command. See .bazelrc for more details.
--config=mkl # Build with MKL support.
--config=mkl_aarch64 # Build with oneDNN and Compute Library for the Arm Architecture (ACL).
--config=monolithic # Config for mostly static monolithic build.
--config=numa # Build with NUMA support.
--config=dynamic_kernels # (Experimental) Build kernels into separate shared objects.
--config=v1 # Build with TensorFlow 1 API instead of TF 2 API.
Preconfigured Bazel build configs to DISABLE default on features:
--config=nogcp # Disable GCP support.
--config=nonccl # Disable NVIDIA NCCL support.

`bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package`
Extracting Bazel installation...
Starting local Bazel server and connecting to it...

**Any other info / logs**
Target //tensorflow/tools/pip_package:build_pip_package failed to build
ERROR: F:/tensorflow/tensorflow/core/framework/BUILD:1400:31 Middleman _middlemen/_S_Stensorflow_Score_Sframework_Cattr_Uvalue_Uproto_Utext-BazelCppSemantics_build_arch_x64_windows-opt failed: (Exit 2): cl.exe failed: error executing command
cd /d C:/users/manyuan/_bazel_manyuan/2qttxlm7/execroot/org_tensorflow
SET INCLUDE=C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include;C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\ucrt;C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\shared;C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\um;C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\winrt;C:\Program Files (x86)\Windows Kits\10\include\10.0.16299.0\cppwinrt
SET PATH=C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\VC\VCPackages;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\TestWindow;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\bin\Roslyn;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\Tools\devinit;C:\Program Files (x86)\Windows Kits\10\bin\10.0.16299.0\x64;C:\Program Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin;C:\Windows\Microsoft.NET\Framework64\v4.0.30319;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\Tools;;C:\Windows\system32
SET PWD=/proc/self/cwd
SET PYTHON_BIN_PATH=D:/JetBrain/python/python.exe
SET PYTHON_LIB_PATH=D:/JetBrain/python/lib/site-packages
SET RUNFILES_MANIFEST_ONLY=1
SET TEMP=C:\Users\manyuan\AppData\Local\Temp
SET TF2_BEHAVIOR=1
SET TMP=C:\Users\manyuan\AppData\Local\Temp
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\cl.exe @bazel-out/x64_windows-opt/bin/tensorflow/core/platform/windows/_objs/platform_port/port.obj.params

**Comment by sushreebarsa** [comment]
@Leemy9612 Please have a look at this [link](https://www.tensorflow.org/install/source_windows), [link1](https://github.com/tensorflow/docs/blob/cbaa96238c823b39e4433c843823895cef903343/site/en/install/source.md#gpu) and verify the configurations.Please let us know if it helps?Thanks!

**Comment by Leemy9612** [comment]
>I want to build tf from source on windows, so I follow the [https://github.com/tensorflow/docs/blob/cbaa96238c823b39e4433c843823895cef903343/site/en/install/source_windows.md](url), but now, the bazel 4.2.1 is too out-of-date to build tf. and it report that `Please upgrade your bazel installation to version 4.2.2 or higher to build TensorFlow!`. So I tried the version 5.0.0 which is also useless.

**Comment by patlevin** [comment]
Building tensorflow on Windows didn't work for me either. I had to apply several workarounds and code changes to make it work (e.g. adding Visual Studio 2022 toolset support to Bazel).

For MSVC 2019 make sure to add a `BAZEL_VC` environment variable that points to `C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC` or your MSBUILD toolkit path, since at least on my machine (also Win11 Pro) Bazel wasn't able to find `cl.exe` either (because in 2022 it still stumbled over a space in the path name🤦‍♂️).

Another issue that arose was the error `realpath not found`, which can be fixed by changing "-c" to "-cl" in [third_party/remote_config/common.bzl:273](https://github.com/tensorflow/tensorflow/blob/master/third_party/remote_config/common.bzl#L273). This [had been fixed before](https://github.com/tensorflow/tensorflow/commit/572f537ea9efba9130e735dfd0feee07bcb44ffe), but was rolled back in later versions for some reason.

In case you get an error that says a directory couldn't be mapped, symlink creation might have failed. This can be fixed by [activating developer mode](https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development) - others might suggest to simply start the `cmd.exe` as administrator instead (which would also work), but I don't recommend that for obvious reasons.
